### ISSUE TITLE
Running auto input on pasted content

### ISSUE THREAD

**Issue by mattberkowitz** [ProseMirror/prosemirror#201]
TL;DR see PR #182 to go straight to code
## Problem

Currently input rules in the auto input module are only executed as a user types. It would be useful for these rules to also run on content being pasted into the document. For example, if straight quotes are converted to curly quotes on typing I would expect that pasting a block of text containing quotes they would also get converted.
#### Event Hooks

The first piece needed to accomplish this is events to hook into with paste. `transformPastedText` and `transformPastedHTML` which allow for a string of the respective to be manipulated prior to being run through the parser. I don't believe these are appropriate for this task because they are not compatible with the `InputRule` callbacks, they would also require parsing and traversing HTML which is something I don't want to really do

An event could be added to manipulate the parsed fragment prior to it being added to the document, which I tried with #177. This also turned out to be incompatible with the InputRule callback format which takes a prosemirror instance as it's first parameter.

The final, and I believe correct, option would be to fire an event _after_ the parsed fragment has been added to the document, with the start and end positions of the insert so that this range can be processed with the InputRules
#### Input Rule RegExp

Currently it is suggested that the RegExp supplied for an input rule end with `$` indicating that the match must be at the end of the tested string. As typing the current blocks content is split at the cursor position so that content after the cursor will not be considered.  In the case of pasted content this will not be the case. Additionally there are potentially multiple matches of a given rule in the pasted content, so the system should ensure that all matches get processed, not just the first
#### Input Rules Only Intended for Typing

It is conceivable that some input rules only be executed on typed content. It's also possible that the inverse is true and some users might wish for a rule to only run on paste. This should be a configurable property of the input rule
## Proposed Solution

I submitted an initial attempt at a solution a few weeks ago (#182). There hasn't yet been any discussion there, hence this issue. I believe this solution would also be useful for things like #90 

As described above, I added an event that fires after pasted content has been added to the document, sending the start and end position of that content as parameters. 

I also created a new class `AutoInputRule` which extends `InputRule`. I wasn't 100% sure of the usage of `InputRules` outside of the autoinput module, so I felt it best to build on top rather than make changes to them. It accepts an options hash as a fourth parameter. It currently only allows for paste parsing to be turned on or off, as turning off type parsing would require changes outside of the autoInput module and it was my intent was to affect the core codebase as little as possible. It also has a new getter `nonEndMatch` which takes the supplied `match` RegExp and returns a new RegExp without the `$` at the end. In a perfect world I think it would work the other way around, the user supplies a general RegExp and a line end version would be generated for use with typing input, but again, I wanted to work within the current paradigm

All `InputRules` defined in the autoInput module, were converted to `AutoInputRules` the ones deal with converting markdown conventions (lists, headings, code, etc) were designated not to run on paste as the markdown parser should handle these.

The autoInput module hooks into the new "pasted" event. It creates a fragment using `sliceBetween` the start and end position arguments. It then loops through each Textblock and InputRule, once a match is found, the callback is called with the corresponding position in the document (as determined by the `docPath` function. A temporary listener is added to the "transform" event in order to update start/end positions and the fragment to be processed as changes are executed. Once a change is found the loops are exited and the whole process is started over, this happen until no matches are found, at which point the "transform" event is unbound and we're good to go!
## Potential Improvements

It might be a good idea to extend the start and end of the processed range to the start and end of the corresponding blocks. Consider an input rule that adds a MarkedRange to the word "blue" so it could be colored blue. If the user had "The US flag is is red, b" and pasted "lue and white". In my current implementation "blue" would not be matched because it is only looking at the actual pasted range

Depending on the larger usage of `InputRule` the separate `AutoInputRule` class might be unnecessary.

If a transform comes from elsewhere (collaborative editing, perhaps) the current process loop should be exited and restarted to account for any potential change in content/position
## Conclusion

The code can definitely be cleaned up some, but it seems to work well. The `domPath` seems a little bootleg, I'm not sure if there's a better way to process a fragment but relate the matches back to the position within the document for `InputRule` callback purposes. The temporary binding to "transform" felt off at first, but I think it works.

Would appreciate feedback/input

**Comment by ericandrewlewis** [comment]
Why do input rules only work with active user input and not pasted content? It looks like this decision was [made early on](https://github.com/ProseMirror/prosemirror/commit/bfc4d2bf) in the project, so there may be some history to the decision that would be useful for this conversation. @marijnh?

**Comment by marijnh** [comment]
Changing text as it is input is an invasive thing to do, and the only reason I think it is a good idea is that, if you're typing, you see it happening, and can intervene. Applying similar inputs to pasted content is much more problematic -- the user is likely not to notice, so instead of a convenient 'text macro' shortcut, we now have insidious mangling of content. Neither Google Docs nor Libre Office apply their smart quote rules to pasted content. I suspect the reasoning behind this is similar.

I do, once I have a representation of pasted content that I'm happy with, intend to expose a hook for user scripts to filter that, maybe even an easy way to apply existing input rules to the text in such content (though as you noticed, the interface used for defining such rules makes that a little awkward). But I don't think this is a generally good idea that should be enabled by default.

**Comment by ericandrewlewis** [comment]
> I do, once I have a representation of pasted content that I'm happy with, intend to expose a hook for user scripts to filter that

Great. For now, perhaps developers can use the generic `transform` event to implement this behavior?

**Comment by marijnh** [comment]
I'm closing this. Input rules solve a specific problem -- auto-replacement during typing -- that is different from the problem of sanitizing or otherwise processing content. The library exposes ways to make it possible to implement such functionality, but it'd look different from input rules.
